language-plutus-core/src/Language/PlutusCore/CBOR.hs

```diff
@@ -0,0 +1,208 @@
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Type
+import           PlutusPrelude
+
+encodeTyBuiltin :: TypeBuiltin -> Encoding
+encodeTyBuiltin bi =
+    let i = case bi of
+            TyByteString -> 0
+            TyInteger    -> 1
+            TySize       -> 2
+    in encodeTag i
+
+decodeTyBuiltin :: Decoder s TypeBuiltin
+decodeTyBuiltin = go =<< decodeTag
+    where go 0 = pure TyByteString
+          go 1 = pure TyInteger
+          go 2 = pure TySize
+          go _ = fail "Failed to decode TypeBuiltin"
+
+encodeBuiltinName :: BuiltinName -> Encoding
+encodeBuiltinName bi =
+    let i = case bi of
+            AddInteger           -> 0
+            SubtractInteger      -> 1
+            MultiplyInteger      -> 2
+            DivideInteger        -> 3
+            RemainderInteger     -> 4
+            LessThanInteger      -> 5
+            LessThanEqInteger    -> 6
+            GreaterThanInteger   -> 7
+            GreaterThanEqInteger -> 8
+            EqInteger            -> 9
+            ResizeInteger        -> 10
+            IntToByteString      -> 11
+            Concatenate          -> 12
+            TakeByteString       -> 13
+            DropByteString       -> 14
+            ResizeByteString     -> 15
+            SHA2                 -> 16
+            SHA3                 -> 17
+            VerifySignature      -> 18
+            EqByteString         -> 19
+            TxHash               -> 20
+            BlockNum             -> 21
+            BlockTime            -> 22
+    in encodeTag i
+
+decodeBuiltinName :: Decoder s BuiltinName
+decodeBuiltinName = go =<< decodeTag
+    where go 0  = pure AddInteger
+          go 1  = pure SubtractInteger
+          go 2  = pure MultiplyInteger
+          go 3  = pure DivideInteger
+          go 4  = pure RemainderInteger
+          go 5  = pure LessThanInteger
+          go 6  = pure LessThanEqInteger
+          go 7  = pure GreaterThanInteger
+          go 8  = pure GreaterThanEqInteger
+          go 9  = pure EqInteger
+          go 10 = pure ResizeInteger
+          go 11 = pure IntToByteString
+          go 12 = pure Concatenate
+          go 13 = pure TakeByteString
+          go 14 = pure DropByteString
+          go 15 = pure ResizeByteString
+          go 16 = pure SHA2
+          go 17 = pure SHA3
+          go 18 = pure VerifySignature
+          go 19 = pure EqByteString
+          go 20 = pure TxHash
+          go 21 = pure BlockNum
+          go 22 = pure BlockTime
+          go _  = fail "Failed to decode BuiltinName"
+
+encodeNatural :: Natural -> Encoding
+encodeNatural = encodeInteger . fromIntegral
+
+decodeNatural :: Decoder s Natural
+decodeNatural = fromIntegral <$> decodeInteger
+
+encodeUnique :: Unique -> Encoding
+encodeUnique (Unique i) = encodeInt i
+
+decodeUnique :: Decoder s Unique
+decodeUnique = Unique <$> decodeInt
+
+-- TODO: should we encode the name or not?
```

Comment:
I think this is an interesting question - if we don't we save space, but if we do then it's actually lossless and much easier to debug. I'd say do encode it until we have reason not to.